File: app/src/main/java/com/wavesplatform/wallet/v1/util/SignUtil.java
/*
 * Created by Eduard Zaydel on 1/4/2019
 * Copyright © 2019 Waves Platform. All rights reserved.
 */

package com.wavesplatform.wallet.v1.util;

import com.google.common.primitives.Bytes;
import com.google.common.primitives.Shorts;
import com.wavesplatform.wallet.v1.crypto.Base58;

public class SignUtil {
    public static byte[] arrayWithSize(String s) throws Base58.InvalidBase58 {
        if (s != null && s.length() > 0) {
            byte[] b = Base58.decode(s);
            return Bytes.concat(Shorts.toByteArray((short) b.length), b);
        } else {
            return Shorts.toByteArray((short) 0);
        }
    }

    public static byte[] arrayOption(String o) throws Base58.InvalidBase58 {
        return org.apache.commons.lang3.StringUtils.isEmpty(o) ?
                new byte[]{0} : Bytes.concat(new byte[]{1}, Base58.decode(o));
    }
}

File: wavesplatform/src/main/java/com/wavesplatform/sdk/utils/SignUtil.kt
/*
 * Created by Eduard Zaydel on 1/4/2019
 * Copyright © 2019 Waves Platform. All rights reserved.
 */

package com.wavesplatform.sdk.utils

import com.google.common.primitives.Bytes
import com.google.common.primitives.Shorts
import com.wavesplatform.sdk.crypto.Base58
import com.wavesplatform.sdk.crypto.WavesCrypto
import java.nio.charset.Charset

object SignUtil {

    @Throws(Base58.InvalidBase58::class)
    fun arrayWithSize(s: String?): ByteArray {
        return if (s != null && s.isNotEmpty()) {
            val b = Base58.decode(s)
            Bytes.concat(Shorts.toByteArray(b.size.toShort()), b)
        } else {
            Shorts.toByteArray(0.toShort())
        }
    }

    @Throws(Base58.InvalidBase58::class)
    fun arrayOption(o: String): ByteArray {
        return if (org.apache.commons.lang3.StringUtils.isEmpty(o))
            byteArrayOf(0)
        else
            Bytes.concat(byteArrayOf(1), Base58.decode(o))
    }


    @Throws(Base58.InvalidBase58::class)
    fun attachmentBytes(attachment: String): ByteArray {
        return if (attachment.isEmpty()) {
            byteArrayOf(0, 0)
        } else {
            WavesCrypto.base58decode(attachment).arrayWithSize()
        }
    }

    @Throws(Base58.InvalidBase58::class)
    fun recipientBytes(recipient: String, version: Byte, chainId: Byte): ByteArray {
        return if (recipient.length <= 30) {
            Bytes.concat(
                byteArrayOf(version),
                byteArrayOf(chainId),
                recipient.parseAlias().toByteArray(Charset.forName("UTF-8")).arrayWithSize()
            )
        } else {
            WavesCrypto.base58decode(recipient)
        }
    }
}
